lib/vrl/stdlib/src/for_each.rs

```diff
@@ -0,0 +1,189 @@
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ForEach;
+
+impl Function for ForEach {
+    fn identifier(&self) -> &'static str {
+        "for_each"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::OBJECT | kind::ARRAY,
+                required: true,
+            },
+            Parameter {
+                keyword: "recursive",
+                kind: kind::BOOLEAN,
+                required: false,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "iterate object",
+                source: r#"count = 0; for_each({ "a": 1, "b": 2 }) -> |_key, value| { count = count + value }; count"#,
+                result: Ok("3"),
+            },
+            Example {
+                title: "recursively iterate object",
+                source: r#"count = 0; for_each({ "a": 1, "b": { "c": 2, "d": 3 } }, recursive: true) -> |_, value| { count = count + (int(value) ?? 0) }; count"#,
+                result: Ok("6"),
+            },
+            Example {
+                title: "iterate array",
+                source: r#"count = 0; for_each([1,2,3]) -> |index, value| { count = count + index + value }; count"#,
+                result: Ok("9"),
+            },
+            Example {
+                title: "recursively iterate array",
+                source: r#"count = 0; for_each([1,2,[3,4]], recursive: true) -> |index, value| { count = count + index + (int(value) ?? 0) }; count"#,
+                result: Ok("14"),
```

Comment:
For the lazy, this expands to:

```
i = 0: 0 + 0 + 1;
i = 1: 1 + 1 + 2;
i = 2: 4 + 2 + 0;
i = 0: 6 + 0 + 3;
i = 1: 9 + 1 + 4;
```

Maybe we can come up with an example that's easier to verify manually? 😄

Another thought that I got here: Wouldn't it be helpful to have a (nesting)`level` argument if `for_each` is called `recursive`? That would help to uniquely identify the position of the iterator from those arguments only

Reply:
Hm, even `level` would not be unique in every case. I could see an argument like `path` that contains the lookup path to the currently iterated value being useful. E.g. the third value would be invoked with `path` equal to `[2][0]`.

If you were iterating over `{ "a": 1, "b": { "c": 2, "d": 3 } }`, then the third value would be invoked with `path` equal to `.b.c`

Reply:
resolved in [9dccfcf](https://github.com/vectordotdev/vector/pull/12382/commits/9dccfcf8e76108e5cf4dbf279e148b6e33aa22f8).